src/relay/pass/util.cc

```diff
@@ -12,122 +12,258 @@
+        ret.push_back(v);
+      }
+    }
+    return ret;
+  }
+
+  Array<TypeVar> CollectBound() {
+    Array<TypeVar> ret;
+    for (const auto& v : bound_type_vars_.data) {
+      ret.push_back(v);
+    }
+    return ret;
+  }
+
+  Array<TypeVar> CollectAll() {
+    Array<TypeVar> ret;
+    for (const auto& v : type_vars_.data) {
+      ret.push_back(v);
+    }
+    return ret;
   }
 
-  Array<TypeVar> Find(const Type& type) {
-    this->VisitType(type);
-    return free_vars_;
+  Array<TypeVar> Free(const Expr& expr) {
+    VisitExpr(expr);
+    return CollectFree();
+  }
+
+  Array<TypeVar> Free(const Type& type) {
+    VisitType(type);
+    return CollectFree();
+  }
+
+  Array<TypeVar> Bound(const Expr& expr) {
+    VisitExpr(expr);
+    return CollectBound();
+  }
+
+  Array<TypeVar> Bound(const Type& type) {
+    VisitType(type);
+    return CollectBound();
+  }
+
+  Array<TypeVar> All(const Expr& expr) {
+    VisitExpr(expr);
+    return CollectAll();
+  }
+
+  Array<TypeVar> All(const Type& type) {
+    VisitType(type);
+    return CollectAll();
   }
 
   void VisitExpr_(const FunctionNode* f) final {
     for (const auto& tp : f->type_params) {
-      bound_vars_.insert(tp);
+      type_vars_.Insert(tp);
+      bound_type_vars_.Insert(tp);
     }
     ExprVisitor::VisitExpr_(f);
   }
 
   void VisitType(const Type& t) final {
-    FreeTypeVarTVisitor(&free_vars_, &bound_vars_)
+    TypeVarTVisitor(&type_vars_, &bound_type_vars_)
         .VisitType(t);
   }
 
  private:
-  // The result list
-  Array<TypeVar> free_vars_;
-  std::unordered_set<TypeVar, NodeHash, NodeEqual> bound_vars_;
+  InsertionSet<TypeVar> type_vars_;
+  InsertionSet<TypeVar> bound_type_vars_;
 };
 
-class FreeVarVisitor : protected ExprVisitor {
+class VarVisitor : protected ExprVisitor {
  public:
-  Array<Var> Find(const Expr& expr) {
+  Array<Var> Free(const Expr& expr) {
     this->VisitExpr(expr);
-    return free_vars_;
+    Array<Var> ret;
+    for (const auto& v : vars_.data) {
+      if (bound_vars_.set.count(v) == 0) {
+        ret.push_back(v);
+      }
+    }
+    return ret;
   }
 
-  void VisitExpr_(const VarNode* var) final {
-    if (bound_vars_.count(var) == 0) {
-      free_vars_.push_back(GetRef<Var>(var));
+  Array<Var> Bound(const Expr& expr) {
+    this->VisitExpr(expr);
+    Array<Var> ret;
+    for (const auto& v : bound_vars_.data) {
+      ret.push_back(v);
     }
+    return ret;
+  }
+
+  Array<Var> All(const Expr& expr) {
+    this->VisitExpr(expr);
+    Array<Var> ret;
+    for (const auto& v : vars_.data) {
+      ret.push_back(v);
+    }
+    return ret;
+  }
+
+  void Bounded(const Var& v) {
+    bound_vars_.Insert(v);
+    vars_.Insert(v);
+  }
+
+  void VisitExpr_(const VarNode* var) final {
+    vars_.Insert(GetRef<Var>(var));
   }
 
   void VisitExpr_(const FunctionNode* op) final {
     for (const auto& param : op->params) {
-      bound_vars_.insert(param.operator->());
+      Bounded(param);
```

Comment:
Is `Bounded` the best name?

Reply:
I guess it should be a verb. I'll change it to `MarkBounded`.